lib/Controller/MessagesController.php

```diff
@@ -437,6 +437,39 @@ public function getSource(int $id): JSONResponse {
 		return $response;
 	}
 
+	/**
+	 * Export a whole message as an .eml file.
+	 *
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 * @TrapError
+	 *
+	 * @param int $id
+	 * @return Response
+	 * @throws ClientException
+	 * @throws ServiceException
+	 */
+	public function export(int $id): Response {
+		try {
+			$message = $this->mailManager->getMessage($this->currentUserId, $id);
+			$mailbox = $this->mailManager->getMailbox($this->currentUserId, $message->getMailboxId());
+			$account = $this->accountService->find($this->currentUserId, $mailbox->getAccountId());
+		} catch (DoesNotExistException $e) {
+			return new JSONResponse([], Http::STATUS_FORBIDDEN);
+		}
+
+		$source = $this->mailManager->getSource(
+			$account,
+			$mailbox->getName(),
+			$message->getUid()
+		);
+		return new AttachmentDownloadResponse(
+			$source,
+			$message->getSubject() . '.eml',
```

Comment:
FYI: This seems to be dangerous at first because the subject might contain illegal characters. I can assure you that this is fine because modern browsers automatically replace illegal characters (e.g. /) which is very sane IMO. I tested latest Chromium and Firefox (Dev Edition).

Reply:
I think it's `OCP\AppFramework\Http\DownloadResponse` that handles the replacement.

https://github.com/nextcloud/server/blob/6209026794962a02d882f30f5f927df81ac22a28/lib/public/AppFramework/Http/DownloadResponse.php#L42